src/c_type_generator.py

```diff
@@ -133,32 +140,46 @@ def union_types(self, a: Optional[CType], b: Optional[CType]) -> Optional[CType]
     def resolve_label(self, sketches: Sketches, node: SkNode) -> SketchNode:
         if isinstance(node, LabelNode):
             self.info("Resolved label: %s", node)
-            return sketches.lookup.get(node.target)
+            return sketches.lookup(node.target)
         return node
 
-    def _succ_no_loadstore(self,
-                           base_dtv: DerivedTypeVariable,
-                           sketches: Sketches,
-                           node: SketchNode,
-                           seen: Set[SketchNode]) -> List[SketchNode]:
+    def _succ_no_loadstore(
+        self,
+        curr_access_path: List[AccessPathLabel],
+        sketches: Sketches,
+        node: SketchNode,
+        seen: Set[SketchNode],
+    ) -> List[Tuple[List[AccessPathLabel], SketchNode]]:
+        """
+        Collect a list of SketchNode successors of `node`
+        in the sketch graph `sketches` by traversing load and store
+        labels. Each `SketchNode` successor is accompanied
+        by the access path that leads to it.
+        The access path will coincide with the node's DTV unless
+        LabelNodes have been traversed.
+        """
         successors = []
         if node not in seen:
             seen.add(node)
             for n in sketches.sketches.successors(node):
+                new_label = sketches.sketches[node][n]["label"]
                 n = self.resolve_label(sketches, n)
                 if n is None:
                     continue
                 if n.dtv.tail in (StoreLabel.instance(), LoadLabel.instance()):
-                    successors.extend(self._succ_no_loadstore(base_dtv, sketches, n, seen))
+                    successors.extend(
+                        self._succ_no_loadstore(
+                            curr_access_path + [new_label], sketches, n, seen
```

Comment:
Is this propagating in the opposite direction? It looks like it's adding a new label instead of removing one.

Reply:
The first argument of `_succ_no_loadstore` now maintains the path traversed so far. For example, if we have a sketch:
`f -> f.load -> f.load.σ4@0`, we will start with a call to `succ_no_loadstore([],sketches,f,{})`, we examine the successors and find one with a label "load" so we have a recursive call `succ_no_loadstore([]+[load], sketches,f.load,{f})`. The recursive call will examine the successors of `f.load`, since those don't have a load or store label, we are at the base case of the recursion and return a list of successor pairs `[([load,σ4@0], f.load.σ4@0)]`. That is the final list returned as the successors of `f`. In this case the access path and the suffix of the successor coincide, but they will not if we traverse edges to label nodes.

Does this clarify your question?

Reply:
I think your version is clearer. Thanks!